### PR TITLE
Initial release of USB gadget mode

### LINKED ISSUE #1: USB gadget mode enabled in the image to allow single wire connection to machine.
Hey, so I found the Pull associated with USB gadget mode.

https://github.com/raspberrypi/linux/pull/1239

Could this be added to the image?

Also looking around the USB gadget mode is supported by Raspberry Pi Zero, Zero W, Model A, Model A+, and PI 4. This would allow you to only have one connection to the Pi and the Pi would have its own private network. 

Setting this up manually would be very difficult but if it could be included in the configurator or at least have the option to edit the config file that would make this process way easier!

See forum for the hard way: [Using RNDIS/Ethernet to connect Raspberry Pi Zero to a computer network.](https://www.mod-bros.com/en/forum/t/using-rndis-ethernet-to-connect-raspberry-pi-zero-to-computer-network~926)

### PR DESCRIPTION
**Pull request by stanoba**
This PR should cover issue #1 

Main changes:
- added new NETWORK_MODE='usb'
- Raspberry is in USB gadget mode configured with static IP address 192.168.7.2
- UUID is now generated from Raspberry's `Serial` number instead of MAC address
- changed config files for USB gadget mode in GoodTFT drivers folder (not yet for Waveshare)
- updated install.sh script
- tested on Raspberry Pi Zero, Zero W and 4  (need to be done on other models)

Note: USB Gadget is compatible with Raspberry Pi 0, 0W, A, A+, and 4 only.  (Pi B, B+, 2, 3, 3A+, and 3B+ lack the required pin for OTG)

### CODE REVIEW

**Comment by stanoba** [comment]
Raspberry Pi Zero W + 3.5" TFT display. It has disabled WiFi and only one cable is attached:

LCD_img1
LCD_img2

**Comment by EnDeRWiGiN** [comment]
Any way I could help test this?

**Comment by stanoba** [comment]
> Any way I could help test this?


Hi @EnDeRWiGiN, I think yes, but you need to update your code on SD card.

Here are steps that you need to follow:
1. Connect Raspberry Pi to internet (needed for installation).

2. Rename folder `mobro-raspberrypi` to `mobro-raspberrypi_old` and `display-drivers` folder to `display-drivers_old`:
```sh
mv mobro-raspberrypi mobro-raspberrypi_old
mv display-drivers display-drivers_old

```

3. Clone my `mobro-raspberrypi` fork and run installation script:
```sh
cd /home/modbros
git clone https://github.com/stanoba/mobro-raspberrypi.git
cd mobro-raspberrypi
chmod +x install.sh
sudo ./install.sh
```
Note: This step is better to do on RPI3 or RPI4 as Zero could be quite slow. After installation you can put SD card back to Zero.

4. Re-connect raspberry to your computer via USB port (not PWR).

5. Wait until boots, you should see new network interface in you computer.

6. Set IP address and net-mask to your new interface e.g. 192.168.7.10/255.255.255.0

7. Open http://192.168.7.2 in your browser and go through wizard.

8. After reboot, everything should be working.

**Comment by EnDeRWiGiN** [comment]
Okay, so I followed your instruction and was able to do steps 1-3 smoothly. There was a problem once I tried to connect over USB. I was getting a "USB could not be deciphered" problem. 

Knowing that it might be a config problem I checked the config.txt file and cmdline.txt. The config file was correct and had "dtoverlay=dwc2" but the cmdline.txt file **did not have** "modules-load=dwc2,g_ether" after rootwait. I added this and was able to proceed.

I was able to do steps 4-8 but afterward, the PI did not see the server even though I could go to the configurator and SSH into it. 

I was able to set the network to private and ran the configuration multiple times but I could not get the Pi to see the modbros Client. I tried the static IP of the computer on that network adapter and the normal Network name but it could not find the client. 

I ended up removing all the firewall rules and reinstalling modbros and I got it to work! (I have a feeling I messed up the firewall rules while tried to get other things working)


TLDR:

- Add "modules-load=dwc2,g_ether" to cmdline.txt file

- Make sure the new network interface has the PI's IP for the gateway. This way it is recognized as a valid network and you can set it to public or private.

- Make sure the modbros application is configured properly with the firewall rules.

- Enjoy! Thank you @stanoba

**Comment by Seraksab** [comment]
Thanks for putting in the work to implement this feature @stanoba 👍 

I'll have a look at all of it once i find time to do so.
But i I guess it will take a bit of time as I want to test it on multiple Pi models and make sure it doesn't interfere with normal network operation (wifi, ethernet) on the models that doesn't support OTG

Also we will need to update all of those Waveshare configuration files.. :/

**Comment by Ezero03** [comment]
Hi, every time I follow this guide I get the exact same log and Windows cant recognice my raspberry with conected USB, althoug modbros app finds it and it works through wifi, not USB.
Attached the installation log, so if there is any error, somebody can help me.

[modbros_ssh_log.txt](https://github.com/ModBros/mobro-raspberrypi/files/5899761/modbros_ssh_log.txt)

**Comment by EnDeRWiGiN** [comment]
@Ezero03 Did you make sure the config.txt file and the cmdline.txt file had correct lines to enable the USB port to become a network adapter?

Did you try the items in my post:  

> Okay, so I followed your instruction and was able to do steps 1-3 smoothly. There was a problem once I tried to connect over USB. I was getting a "USB could not be deciphered" problem.
> 
> Knowing that it might be a config problem I checked the config.txt file and cmdline.txt. The config file was correct and had "dtoverlay=dwc2" but the cmdline.txt file **did not have** "modules-load=dwc2,g_ether" after rootwait. I added this and was able to proceed.
> 
> I was able to do steps 4-8 but afterward, the PI did not see the server even though I could go to the configurator and SSH into it.
> 
> I was able to set the network to private and ran the configuration multiple times but I could not get the Pi to see the modbros Client. I tried the static IP of the computer on that network adapter and the normal Network name but it could not find the client.
> 
> I ended up removing all the firewall rules and reinstalling modbros and I got it to work! (I have a feeling I messed up the firewall rules while tried to get other things working)
> 
> TLDR:
> 
> * Add "modules-load=dwc2,g_ether" to cmdline.txt file
> * Make sure the new network interface has the PI's IP for the gateway. This way it is recognized as a valid network and you can set it to public or private.
> * Make sure the modbros application is configured properly with the firewall rules.
> * Enjoy! Thank you stanoba
